ballista/rust/scheduler/src/state/execution_graph.rs

```diff
@@ -581,6 +726,54 @@ impl ExecutionGraph {
         }
     }
 
+    /// Convert running stage to be unresolved
+    fn rollback_running_stage(&mut self, stage_id: usize) -> Result<bool> {
```

Comment:
I'm a little confused by this. If we have tasks that are already running then they will either fail (if they cannot fetch their input partitions from the lost executor) or have already fetched the partitions from the lost executor in which case we shouldn't roll back. I guess I'm not clear on how this process interacts with the incoming task updates. Shouldn't the rollback be in response to task failure instead of just the executor missing a heartbeat?

Reply:
Yes, in this PR, If it decides to rollback, the enter stage will be rollbacked and rescheduled.  Even there are completed tasks, they will be rollback and rescheduled. This approach is simple and safe.  Because its input stages need to be recomputed, we are not sure whether the re-computation result is deterministic or not.

Regarding when to trigger the executor lost handling, as I noted in the PR's description, there could be four cases.  I will implement the shuffle reader failure check in a seperate PR which will focus on the task failure handing case.

1) Executor graceful shutdown, Executor will call RPC to notify the Scheduler it is stopped.
2) Scheduler expired the dead Executors that do not receive heartbeats from them.
3) Shuffle Reader failure due to connectivity issue in reduce tasks(Not implemented, will  cover this in following task failure related PRs).
4) Grpc connection lost detect(Not implemented).

I think we still need the Scheduler to expire the dead executors actively. For example if we have heavy tasks in the map stage and the executor is lost, no task updates come back from the lost executor, in this case we can only rely on the Scheduler expiration logic.
 
We can have the expiration time longer than the executor's heartbeat time interval, so only after missing two or three consecutive heartbeats, the Scheduler consider it is dead.

Reply:
> We can have the expiration time longer than the executor's heartbeat time interval, so only after missing two or three consecutive heartbeats, the Scheduler consider it is dead.

I think this is a good idea. We probably want to be pretty conservative in marking an executor dead due to missing heartbeats.